homeassistant/components/azure_data_explorer/__init__.py

```diff
@@ -0,0 +1,230 @@
+        if self._next_send_remover:
+            self._next_send_remover()
+        if self._listener_remover:
+            self._listener_remover()
+        await self._queue.put(
+            (3, (utcnow(), None))
+        )  # <---WHY IS THIS NEEDED!!! Creates error when sending none
+        await self.async_send(None)
+        await self._queue.join()
+
+    def update_options(self, new_options: dict[str, Any]) -> None:
+        """Update options."""
+        self._send_interval = new_options[CONF_SEND_INTERVAL]
+
+    async def test_connection(self) -> None:
+        """Test the connection to the Azure Data Explorer service."""
+        await self.hass.async_add_executor_job(self._client.test_connection)
+
+        return None
+
+    def _schedule_next_send(self) -> None:
+        """Schedule the next send."""
+        if not self._shutdown:
+            self._next_send_remover = async_call_later(
+                self.hass, self._send_interval, self.async_send
+            )
+
+    async def async_listen(self, event: Event) -> None:
+        """Listen for new messages on the bus and queue them for AEH."""
+        if state := event.data.get("new_state"):
+            await self._queue.put((2, (event.time_fired, state)))
+
+    async def async_send(self, _) -> None:
+        """Write preprocessed events to Azure Data Explorer."""
+
+        adx_events = ""
+        while not self._queue.empty():
+            _, event = self._queue.get_nowait()
+            dropped = 0
+            adx_event, dropped = self._parse_event(*event, dropped)
+            if dropped == 0:
+                adx_events += str(adx_event)
+
+        if adx_events != "":
+
+            try:
+                await self.hass.async_add_executor_job(
+                    lambda: self._client.ingest_data(adx_events)
+                )
+            except Exception as err:  # pylint: disable=broad-except
+                _LOGGER.error(err)
+
+        self._schedule_next_send()
+
+    def _parse_event(
+        self,
+        time_fired: datetime,
+        state: State | None,
+        dropped: int,
+    ) -> tuple[str | None, int]:
+        """Parse event by checking if it needs to be sent, and format it."""
+        self._queue.task_done()
+        if not state:
+            self._shutdown = True
+            return None, dropped + 1
+        if state.state in FILTER_STATES or not self._entities_filter(state.entity_id):
+            return None, dropped + 1
+        if (utcnow() - time_fired).seconds > self._max_delay + self._send_interval:
+            return None, dropped + 1
+        try:
+            json_string = bytes(json.dumps(obj=state, cls=JSONEncoder).encode("utf-8"))
+            json_dictionary = json.loads(json_string)
+            json_event = json.dumps(json_dictionary)
+        except Exception as exp:  # pylint: disable=broad-except
+            _LOGGER.error(exp)
+            return ("", 1)
```

Comment:
This should be dropped + 1, and what does an empty string mean? and what kinds of errors could be possible here, no way to catch those more specifically?

Reply:
Changed, I was not advare of the meaning of the dropped counter.